### PR TITLE
Use edts media time offset if edits list duration is entire media duration.

### PR DESCRIPTION
**Pull request by mayurk2**
It will fix the AV sync issue mentioned here - https://github.com/google/ExoPlayer/issues/10996 

(Test media file: https://drive.google.com/file/d/156XrRCy0Nxk6_u1fgQa9JVEExy3JeLRk/view?usp=sharing)

### CODE REVIEW

**Comment by tonihei** [comment]
Thanks for trying to fix this! Could you open the PR against the [main branch of the AndroidX media3 library](https://github.com/androidx/media/tree/main)? This makes it a lot easier for us to merge the change. I'll leave some comment in this PR still for your review.

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java [hunk 1]
```diff
@@ -968,6 +970,27 @@ private static long parseTfdt(ParsableByteArray tfdt) {
     return version == 1 ? tfdt.readUnsignedLongToLong() : tfdt.readUnsignedInt();
   }
 
+  /**
+   * Retunrs if edts list duration is for the entire media timeline.
```
Comment (tonihei) [inline library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java]: There is no need to document a private method with Javadoc as long as the method name is self-explanatory (which it is).

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java [hunk 2]
```diff
@@ -968,6 +970,27 @@ private static long parseTfdt(ParsableByteArray tfdt) {
...
+    // Currently we only support a single edit that moves the entire media timeline (indicated by
+    // duration == 0 or duration == track duration).
+    // Other uses of edit lists are uncommon and unsupported.
+    if (track.editListDurations != null && track.editListDurations.length == 1) {
+      long edtsDurationUs = Util.scaleLargeTimestamp(track.editListDurations[0],
+          C.MICROS_PER_SECOND, track.movieTimescale);
+      return edtsDurationUs == 0
+          || (track.durationUs > MICROS_PER_MILLI
+          && edtsDurationUs >= track.durationUs - MICROS_PER_MILLI
+          && edtsDurationUs <= track.durationUs + MICROS_PER_MILLI);
```
Comment (tonihei) [inline library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java]: The `edtsListDuration` is the duration of the edit itself. So the end time of the edit is `editListMediaTime + editListDuration`. And I think if we want to check if the edit list covers the entire segment, we can accept all edits that exceed the track duration. So the condition can just be:
`scaleToUs(editListMediaTime + editListDuration) >= track.durationUs`

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java (hunk 1 above)
Comment (mayurk2) [reply library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java]: Updated @tonihei .

File: library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java (hunk 2 above)
Comment (mayurk2) [reply library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java]: Agreed. Updated the code @tonihei .

**Comment by tonihei** [comment]
Thanks for the updates - did you see my comment above about opening this PR against the main branch of media3?

**Comment by mayurk2** [comment]
> Thanks for trying to fix this! Could you open the PR against the [main branch of the AndroidX media3 library](https://github.com/androidx/media/tree/main)? This makes it a lot easier for us to merge the change. I'll leave some comment in this PR still for your review.

@tonihei Opened a PR - https://github.com/androidx/media/pull/255

**Comment by mayurk2** [comment]
@tonihei should we merge this PR too? Or there are no more releases (minor/major) for Exoplayer?

**Comment by mayurk2** [comment]
@tonihei should I raise PR to dev-v2 branch? Please let me know.
Opened it anyway :)
https://github.com/google/ExoPlayer/pull/11017

**Comment by tonihei** [comment]
The media3 one is enough, thanks. We have a single codebase (in media3 package names) internally and the code gets pushed to both repositories automatically.

**Pull request closed without merging**